woocommerce/woocommerce
# dev: add core profiler state machine tests

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Added a introductory setup for testing the core profiler state machine and views. 

This setup is configured by a set of view assertions that match with each core profiler page that should be displayed to the user, along with the DOM events that should result in reaching these pages.

It covers arriving at every possible state through the events that are algorithmically determined to be required to reach those states.

There are also assertions to assert each page's Jest snapshot, but I'm not too sure if it adds much to the tests. Get the feeling it might be a bit noisy when any changes are made.

Closes #37995 



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Clone this repo and run `pnpm install` (because there's a new package @xstate/test)
2. Run `pnpm run --filter woocommerce/client/admin test:watch -- ./client/core-profiler/test/core-profiler-machine.test.tsx`
3. Test should run.

Should look like:
<img width="548" alt="image" src="https://user-images.githubusercontent.com/27843274/236750601-e57be0dd-0ed8-42cf-9de6-ec84e8c86eb7.png">
